File: API/fikrfree-client-sdk.js
/**
 * FikrFree Assistant API Client SDK (JavaScript)
 * A simple JavaScript SDK for integrating with FikrFree Assistant API
 * 
 * Usage:
 *   // Node.js
 *   const FikrFreeClient = require('./fikrfree-client-sdk');
 *   
 *   // Browser (ES6 modules)
 *   import FikrFreeClient from './fikrfree-client-sdk.js';
 *   
 *   // Basic usage
 *   const client = new FikrFreeClient('https://your-domain.com/api/v1');
 *   const response = await client.chat('What are BIMA insurance plans?');
 *   console.log(response.response);
 */

class FikrFreeAPIError extends Error {
    constructor(message, statusCode = null) {
        super(message);
        this.name = 'FikrFreeAPIError';
        this.statusCode = statusCode;
    }
}

class FikrFreeClient {
    /**
     * Initialize the FikrFree API client
     * @param {string} baseUrl - Base URL of the FikrFree API
     */
    constructor(baseUrl = 'http://localhost:8000/api/v1') {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.sessionId = null;
        this.defaultHeaders = {
            'Content-Type': 'application/json',
            'User-Agent': 'FikrFree-JavaScript-SDK/1.0'
        };
    }

    /**
     * Make HTTP request to API
     * @private
     */
    async _makeRequest(method, endpoint, data = null) {
        const url = `${this.baseUrl}${endpoint}`;
        const options = {
            method: method.toUpperCase(),
            headers: { ...this.defaultHeaders }
        };

        if (data) {
            options.body = JSON.stringify(data);
        }

        try {
            const response = await fetch(url, options);
            
            if (!response.ok) {
                let errorMessage = `API request failed with status ${response.status}`;
                try {
                    const errorData = await response.json();
                    errorMessage += `: ${errorData.message || response.statusText}`;
                } catch {
                    errorMessage += `: ${response.statusText}`;
                }
                throw new FikrFreeAPIError(errorMessage, response.status);
            }

            return await response.json();
        } catch (error) {
            if (error instanceof FikrFreeAPIError) {
                throw error;
            }
            throw new FikrFreeAPIError(`Network error: ${error.message}`);
        }
    }

    /**
     * Check API health status
     * @returns {Promise<Object>} Health status information
     */
    async healthCheck() {
        return await this._makeRequest('GET', '/health');
    }

    /**
     * Create a new chat session
     * @returns {Promise<string>} Session ID
     */
    async createSession() {
        const response = await this._makeRequest('POST', '/sessions/start');
        this.sessionId = response.session_id;
        return this.sessionId;
    }

    /**
     * Send a message to the chatbot
     * @param {string} message - Message to send
     * @param {boolean} stream - Whether to use streaming response
     * @param {boolean} autoCreateSession - Automatically create session if none exists
     * @returns {Promise<Object>} Chat response
     */
    async chat(message, stream = false, autoCreateSession = true) {
        if (!message || !message.trim()) {
            throw new Error('Message cannot be empty');
        }

        // Create session if needed
        if (!this.sessionId && autoCreateSession) {
            await this.createSession();
        } else if (!this.sessionId) {
            throw new FikrFreeAPIError('No active session. Call createSession() first or set autoCreateSession=true');
        }

        const data = { message: message.trim(), stream };

        if (stream) {
            return await this._handleStreamingResponse(data);
        } else {
            return await this._makeRequest('POST', `/sessions/${this.sessionId}/chat`, data);
        }
    }

    /**
     * Handle streaming response
     * @private
     */
    async _handleStreamingResponse(data) {
        const url = `${this.baseUrl}/sessions/${this.sessionId}/chat`;
        const response = await fetch(url, {
            method: 'POST',
            headers: { ...this.defaultHeaders },
            body: JSON.stringify(data)
        });

        if (!response.ok) {
            throw new FikrFreeAPIError(`Streaming request failed: ${response.statusText}`);
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let fullResponse = '';
        let finalData = null;

        const chunks = [];

        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                const chunk = decoder.decode(value);
                const lines = chunk.split('\n');

                for (const line of lines) {
                    if (line.startsWith('data: ')) {
                        try {
                            const eventData = JSON.parse(line.slice(6));
                            
                            if (eventData.type === 'content') {
                                fullResponse += eventData.chunk;
                                chunks.push(eventData.chunk);
                                
                                // Emit chunk event if callback is provided
                                if (this.onChunk) {
                                    this.onChunk(eventData.chunk);
                                }
                            } else if (eventData.type === 'complete') {
                                finalData = eventData;
                                break;
                            } else if (eventData.type === 'error') {
                                throw new FikrFreeAPIError(`Streaming error: ${eventData.message}`);
                            }
                        } catch (parseError) {
                            // Skip invalid JSON lines
                            continue;
                        }
                    }
                }

                if (finalData) break;
            }
        } finally {
            reader.releaseLock();
        }

        return {
            ...finalData,
            chunks,
            full_response: fullResponse
        };
    }

    /**
     * Set callback for streaming chunks
     * @param {function} callback - Function to call for each chunk
     */
    onChunk(callback) {
        this.onChunk = callback;
    }

    /**
     * Get conversation history for current session
     * @returns {Promise<Array>} Array of message objects
     */
    async getHistory() {
        if (!this.sessionId) {
            throw new FikrFreeAPIError('No active session');
        }

        const response = await this._makeRequest('GET', `/sessions/${this.sessionId}/history`);
        return response.messages;
    }

    /**
     * Get information about current session
     * @returns {Promise<Object>} Session information
     */
    async getSessionInfo() {
        if (!this.sessionId) {
            throw new FikrFreeAPIError('No active session');
        }

        return await this._makeRequest('GET', `/sessions/${this.sessionId}/info`);
    }

    /**
     * Delete the current session
     * @returns {Promise<boolean>} True if session was deleted successfully
     */
    async deleteSession() {
        if (!this.sessionId) {
            return true; // No session to delete
        }

        try {
            await this._makeRequest('DELETE', `/sessions/${this.sessionId}`);
            this.sessionId = null;
            return true;
        } catch (error) {
            return false;
        }
    }

    /**
     * Get API statistics
     * @returns {Promise<Object>} API statistics
     */
    async getStats() {
        return await this._makeRequest('GET', '/sessions/stats');
    }
}

// Convenience functions for quick usage

/**
 * Quick one-off chat without session management
 * @param {string} message - Message to send
 * @param {string} apiUrl - API base URL
 * @returns {Promise<string>} Bot's response as string
 */
async function quickChat(message, apiUrl = 'http://localhost:8000/api/v1') {
    const client = new FikrFreeClient(apiUrl);
    try {
        const response = await client.chat(message);
        return response.response;
    } finally {
        await client.deleteSession();
    }
}

/**
 * Send multiple messages in a single session
 * @param {Array<string>} messages - Array of messages to send
 * @param {string} apiUrl - API base URL
 * @returns {Promise<Array>} Array of chat responses
 */
async function batchChat(messages, apiUrl = 'http://localhost:8000/api/v1') {
    const client = new FikrFreeClient(apiUrl);
    const responses = [];
    
    try {
        for (const message of messages) {
            const response = await client.chat(message);
            responses.push(response);
        }
        return responses;
    } finally {
        await client.deleteSession();
    }
}

// Example usage and testing
async function demonstrateSDK() {
    console.log('=== FikrFree JavaScript SDK Demonstration ===\n');

    try {
        // Example 1: Simple usage
        console.log('=== Example 1: Simple Chat ===');
        const response = await quickChat('What are BIMA insurance plans?');
        console.log(`Bot: ${response.substring(0, 100)}...`);

        console.log('\n=== Example 2: Session-based Chat ===');
        const client = new FikrFreeClient();
        
        // Check API health
        const health = await client.healthCheck();
        console.log(`API Status: ${health.status}`);

        // Chat with context
        const questions = [
            'What insurance plans do you offer?',
            'Tell me more about BIMA Bronze plan',
            'What\'s the price?',
            'BIMA ke plans ke bare mein batao' // Roman Urdu
        ];

        for (const question of questions) {
            console.log(`\nUser: ${question}`);
            const response = await client.chat(question);
            console.log(`Bot (${response.language_detected}): ${response.response.substring(0, 80)}...`);
        }

        // Show conversation history
        console.log('\n=== Conversation History ===');
        const history = await client.getHistory();
        console.log(`Total messages: ${history.length}`);

        // Show last 4 messages
        const recentMessages = history.slice(-4);
        recentMessages.forEach((msg, i) => {
            const role = msg.role === 'user' ? 'You' : 'Bot';
            console.log(`${i + 1}. ${role}: ${msg.content.substring(0, 50)}...`);
        });

        // Clean up
        await client.deleteSession();

        console.log('\n=== Example 3: Streaming Chat ===');
        const streamClient = new FikrFreeClient();
        
        // Set up chunk handler
        let streamedText = '';
        streamClient.onChunk = (chunk) => {
            streamedText += chunk;
            process.stdout.write(chunk); // Show real-time typing
        };

        console.log('User: Tell me about your services (streaming)');
        console.log('Bot: ');
        
        const streamResponse = await streamClient.chat('Tell me about your services', true);
        console.log(`\n[Stream complete - Language: ${streamResponse.language_detected}]`);
        
        await streamClient.deleteSession();

        console.log('\n=== Example 4: Batch Processing ===');
        const batchQuestions = [
            'What is FikrFree?',
            'How do I file a claim?',
            'What are your contact details?'
        ];

        const batchResponses = await batchChat(batchQuestions);
        batchResponses.forEach((response, i) => {
            console.log(`${i + 1}. Language: ${response.language_detected}`);
            console.log(`   Response: ${response.response.substring(0, 60)}...`);
        });

    } catch (error) {
        console.error('Error:', error.message);
    }

    console.log('\n=== SDK demonstration complete! 🎉 ===');
}

// Export for different environments
if (typeof module !== 'undefined' && module.exports) {
    // Node.js environment
    module.exports = {
        FikrFreeClient,
        FikrFreeAPIError,
        quickChat,
        batchChat,
        demonstrateSDK
    };
} else if (typeof window !== 'undefined') {
    // Browser environment
    window.FikrFreeClient = FikrFreeClient;
    window.FikrFreeAPIError = FikrFreeAPIError;
    window.quickChat = quickChat;
    window.batchChat = batchChat;
}

// Auto-run demonstration if this file is executed directly
if (typeof require !== 'undefined' && require.main === module) {
    // Check if fetch is available (Node.js 18+ or with polyfill)
    if (typeof fetch === 'undefined') {
        console.log('This example requires Node.js 18+ or a fetch polyfill.');
        console.log('To install fetch polyfill: npm install node-fetch');
        console.log('Then add: global.fetch = require("node-fetch");');
    } else {
        demonstrateSDK().catch(console.error);
    }
}

File: API/fikrfree_client_sdk.py
"""
FikrFree Assistant API Client SDK
A simple Python SDK for integrating with FikrFree Assistant API
"""

import requests
import json
from typing import Optional, Dict, List
from dataclasses import dataclass
from datetime import datetime


@dataclass
class ChatResponse:
    """Response from chat API"""
    session_id: str
    response: str
    language_detected: str
    timestamp: str
    message_count: int


@dataclass
class SessionInfo:
    """Session information"""
    session_id: str
    created_at: str
    last_activity: str
    message_count: int
    is_active: bool


@dataclass
class Message:
    """Chat message"""
    role: str
    content: str
    timestamp: str


class FikrFreeAPIError(Exception):
    """Custom exception for API errors"""
    pass


class FikrFreeClient:
    """
    FikrFree Assistant API Client
    
    A simple, easy-to-use client for the FikrFree Assistant API.
    Handles session management, chat interactions, and error handling.
    
    Example:
        client = FikrFreeClient("https://your-domain.com/api/v1")
        
        # Simple chat
        response = client.chat("What are BIMA insurance plans?")
        print(response.response)
        
        # Multiple messages in same session
        client.chat("Tell me more about Bronze plan")
        client.chat("What about Silver plan?")
        
        # Get conversation history
        history = client.get_history()
    """
    
    def __init__(self, base_url: str = "http://localhost:8000/api/v1"):
        """
        Initialize the FikrFree API client
        
        Args:
            base_url: Base URL of the FikrFree API
        """
        self.base_url = base_url.rstrip('/')
        self.session_id: Optional[str] = None
        self.session = requests.Session()
        
        # Set default headers
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'FikrFree-Python-SDK/1.0'
        })
    
    def _make_request(self, method: str, endpoint: str, data: Optional[Dict] = None) -> Dict:
        """Make HTTP request to API"""
        url = f"{self.base_url}{endpoint}"
        
        try:
            if method.upper() == 'GET':
                response = self.session.get(url)
            elif method.upper() == 'POST':
                response = self.session.post(url, json=data)
            elif method.upper() == 'DELETE':
                response = self.session.delete(url)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")
            
            # Check for HTTP errors
            if response.status_code >= 400:
                error_msg = f"API request failed with status {response.status_code}"
                try:
                    error_data = response.json()
                    error_msg += f": {error_data.get('message', response.text)}"
                except:
                    error_msg += f": {response.text}"
                raise FikrFreeAPIError(error_msg)
            
            return response.json()
            
        except requests.exceptions.RequestException as e:
            raise FikrFreeAPIError(f"Network error: {str(e)}")
    
    def health_check(self) -> Dict:
        """
        Check API health status
        
        Returns:
            Dict with health status information
        """
        return self._make_request('GET', '/health')
    
    def create_session(self) -> str:
        """
        Create a new chat session
        
        Returns:
            Session ID string
        """
        response = self._make_request('POST', '/sessions/start')
        self.session_id = response['session_id']
        return self.session_id
    
    def chat(self, message: str, stream: bool = False, auto_create_session: bool = True) -> ChatResponse:
        """
        Send a message to the chatbot
        
        Args:
            message: Message to send
            stream: Whether to use streaming response (not implemented in this SDK)
            auto_create_session: Automatically create session if none exists
        
        Returns:
            ChatResponse object with bot's reply
        """
        if not message.strip():
            raise ValueError("Message cannot be empty")
        
        # Create session if needed
        if not self.session_id and auto_create_session:
            self.create_session()
        elif not self.session_id:
            raise FikrFreeAPIError("No active session. Call create_session() first or set auto_create_session=True")
        
        # Send message
        data = {"message": message, "stream": stream}
        response = self._make_request('POST', f'/sessions/{self.session_id}/chat', data)
        
        return ChatResponse(
            session_id=response['session_id'],
            response=response['response'],
            language_detected=response['language_detected'],
            timestamp=response['timestamp'],
            message_count=response['message_count']
        )
    
    def get_history(self) -> List[Message]:
        """
        Get conversation history for current session
        
        Returns:
            List of Message objects
        """
        if not self.session_id:
            raise FikrFreeAPIError("No active session")
        
        response = self._make_request('GET', f'/sessions/{self.session_id}/history')
        
        return [
            Message(
                role=msg['role'],
                content=msg['content'],
                timestamp=msg['timestamp']
            )
            for msg in response['messages']
        ]
    
    def get_session_info(self) -> SessionInfo:
        """
        Get information about current session
        
        Returns:
            SessionInfo object
        """
        if not self.session_id:
            raise FikrFreeAPIError("No active session")
        
        response = self._make_request('GET', f'/sessions/{self.session_id}/info')
        
        return SessionInfo(
            session_id=response['session_id'],
            created_at=response['created_at'],
            last_activity=response['last_activity'],
            message_count=response['message_count'],
            is_active=response['is_active']
        )
    
    def delete_session(self) -> bool:
        """
        Delete the current session
        
        Returns:
            True if session was deleted successfully
        """
        if not self.session_id:
            return True  # No session to delete
        
        try:
            self._make_request('DELETE', f'/sessions/{self.session_id}')
            self.session_id = None
            return True
        except FikrFreeAPIError:
            return False
    
    def get_stats(self) -> Dict:
        """
        Get API statistics (requires no session)
        
        Returns:
            Dict with API statistics
        """
        return self._make_request('GET', '/sessions/stats')
    
    def __enter__(self):
        """Context manager entry"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit - clean up session"""
        self.delete_session()


# Convenience functions for quick usage
def quick_chat(message: str, api_url: str = "http://localhost:8000/api/v1") -> str:
    """
    Quick one-off chat without session management
    
    Args:
        message: Message to send
        api_url: API base URL
    
    Returns:
        Bot's response as string
    """
    with FikrFreeClient(api_url) as client:
        response = client.chat(message)
        return response.response


def batch_chat(messages: List[str], api_url: str = "http://localhost:8000/api/v1") -> List[ChatResponse]:
    """
    Send multiple messages in a single session
    
    Args:
        messages: List of messages to send
        api_url: API base URL
    
    Returns:
        List of ChatResponse objects
    """
    with FikrFreeClient(api_url) as client:
        responses = []
        for message in messages:
            response = client.chat(message)
            responses.append(response)
        return responses


# Example usage
if __name__ == "__main__":
    # Example 1: Simple usage
    print("=== Example 1: Simple Chat ===")
    response = quick_chat("What are BIMA insurance plans?")
    print(f"Bot: {response}")
    
    print("\n=== Example 2: Session-based Chat ===")
    with FikrFreeClient() as client:
        # Check API health
        health = client.health_check()
        print(f"API Status: {health['status']}")
        
        # Chat with context
        questions = [
            "What insurance plans do you offer?",
            "Tell me more about BIMA Bronze plan",
            "What's the price?",
            "BIMA ke plans ke bare mein batao"  # Roman Urdu
        ]
        
        for question in questions:
            print(f"\nUser: {question}")
            response = client.chat(question)
            print(f"Bot ({response.language_detected}): {response.response[:100]}...")
        
        # Show conversation history
        print(f"\n=== Conversation History ===")
        history = client.get_history()
        print(f"Total messages: {len(history)}")
        
        for i, msg in enumerate(history[-4:], 1):  # Show last 4 messages
            role = "You" if msg.role == "user" else "Bot"
            print(f"{i}. {role}: {msg.content[:50]}...")
    
    print("\n=== Example 3: Batch Processing ===")
    questions = [
        "What is FikrFree?",
        "How do I file a claim?",
        "What are your contact details?"
    ]
    
    responses = batch_chat(questions)
    for i, response in enumerate(responses, 1):
        print(f"{i}. Language: {response.language_detected}")
        print(f"   Response: {response.response[:80]}...")
    
    print("\n=== Example 4: Error Handling ===")
    try:
        # This will fail with invalid URL
        client = FikrFreeClient("http://invalid-url")
        client.chat("Hello")
    except FikrFreeAPIError as e:
        print(f"API Error: {e}")
    
    print("\nSDK demonstration complete! 🎉")